docs/design_notes/type_promotion.md

```diff
@@ -0,0 +1,722 @@
+
+Numpy's choice here is to promote to `float64`:
+
+```{code-cell}
+:id: ZKUCX9ryjzhN
+
+(np.uint64(1) + np.int64(1)).dtype
+```
+
++++ {"id": "yV-9Uka-j98S"}
+
+However, this may be a surprising convention: it's the only case in which you add two integers and don't get an integer out!
+For now, we will leave `uint64` promotion undefined, and return to it later.
+
++++ {"id": "BBNC59v_pTSY"}
+
+## Mixed Promotion: Integer and Floating
+
+When promoting integers to floating point, we might start with the same thought process as mixed promotion between signed and unsigned integers. A 16-bit signed or unsigned integer cannot be represented at full precision by a 16-bit float, which has only 10 bits of mantissa. Therefore, it might make sense to promote integers to floats represented by twice the number of bits:
+
+```{code-cell}
+:cellView: form
+:id: GT5uPYlMs3sw
+:tags: [hide-input]
+
+#@title
+import networkx as nx
+import matplotlib.pyplot as plt
+lattice = {
+  'i*': ['f*', 'u8', 'i8'], 'f*': ['c*', 'f16'], 'c*': ['c64'],
+  'u8': ['u16', 'i16', 'f16'], 'u16': ['u32', 'i32', 'f32'], 'u32': ['u64', 'i64', 'f64'],
+  'i8': ['i16', 'f16'], 'i16': ['i32', 'f32'], 'i32': ['i64', 'f64'],
+  'f16': ['f32'], 'f32': ['f64', 'c64'], 'f64': ['c128'],
+  'c64': ['c128']
+}
+graph = nx.from_dict_of_lists(lattice, create_using=nx.DiGraph)
+pos = {
+  'i*': [-1.25, 0.5], 'f*': [-0.5, 2], 'c*': [0, 3],
+  'u8': [0.5, 0], 'u16': [1.5, 0], 'u32': [2.5, 0], 'u64': [3.5, 0],
+  'i8': [0, 1], 'i16': [1, 1], 'i32': [2, 1], 'i64': [3, 1],
+  'f16': [0.5, 2], 'f32': [1.5, 2], 'f64': [2.5, 2],
+  'c64': [2, 3], 'c128': [3, 3],
+}
+fig, ax = plt.subplots(figsize=(6, 5))
+nx.draw(graph, with_labels=True, node_size=1500, node_color='lightgray', pos=pos, ax=ax)
+```
+
++++ {"id": "gYIJaqkCuh35"}
+
+This is effectively what Numpy type promotion does, but in doing so it breaks the lattice property of the graph: for example, the pair *{i8, u8}* no longer has a unique least upper bound: the possibilities are *i16* and *f16*, which are unorderable on the graph. This turns out to be the source of NumPy's non-associative type promotion highlighted above.
+
+Can we come up with a modification of NumPy's promotion rules, such that it will satisfy the lattice property, while also giving sensible results for mixed type promotion? There are a few approaches we could take here.
+
++++ {"id": "nLXKOk48lfY2"}
+
+### Option 0: Leave integer/floating mixed precision undefined
+
+To make behavior utterly predictable (at some cost to user convenience), a defensible choice would be to leave mixed integer/float type promotion undefined, stopping with the partial lattice from the previous section. The cost would be requiring the user to explicitly type-cast when operating between integer and floating-point quantities.
```

Comment:
One thing to note: the previous lattice would still allow Python scalars to be promoted to floats.

Reply:
Done.